src/providers/oracle/qgsoracleprovider.cpp

```diff
@@ -2614,6 +2614,15 @@ bool QgsOracleProvider::getGeometryDetails()
     return true;
   }
 
+  // Do not re-check geometry type if it is already known (f.e. when re-loading a project)
+  if ( mRequestedGeomType != QgsWkbTypes::Unknown && mSrid > 0 )
```

Comment:
IMPORTANT: I'm not an expert/developer.
My impression is this code is executed once when we create or re-open an oracle layer (when opening a project). Not for a specific feature, but for the whole layer.
_mDetectedGeomType_ is the type of geometry we've already chosen/found when creating a layer. So styling, labelling etc. is already set-up according to this geometry type.
When re-opening the layer again, this routine re-checks the type of geometry (and in oracle, which does not hold geometry type in meta-tables, this is an expensive action). And it is hard to say what would happen if geometry type **would** change say from polygon to point... all symbolisation/labelling settings will be partly incorrect/irrelevant.
If I'm not mistaken, I'm overwriting _mDetectedGeomType_ with type which was found in previous sessions. So if a user has added a layer with type polygon, I would use polygon, if a user added a layer with line, I would use line etc.

Can you think of some use case/test I could try to test your version?

P.S. Regarding https://github.com/qgis/QGIS/pull/34358 we actually are very happy that QGIS ignores the difference between polygon and multipolygon as these are all polygons in general (same symbolisation etc.).

Reply:
There is one provider for each layer, created when you create the layer (open the project for instance). 

mDetectedGeomType is the geometry detected from the database, mRequestedGeomType is the one specified by the user (specified in uri you could see on tooltip).

> Can you think of some use case/test I could try to test your version?

You can have a table with mixed polygons and multipolygons and want to display only polygons. That's the way other providers work in QGIS, and that's the desired fix of the PR #34358. 

> P.S. Regarding #34358 we actually are very happy that QGIS ignores the difference between polygon and multipolygon as these are all polygons in general (same symbolisation etc.).

I understand your issue here. @SebastienPeillet It will indeed break the workflow of people relying on the fact that geomtype and multigeomtype are mixed now but I don't see a way to not break this with your PR. My guess is that you should have either 2 layers or convert to multi the single geometry with a view.

Reply:
>> Can you think of some use case/test I could try to test your version?
> You can have a table with mixed polygons and multipolygons and want
> to display only polygons.

  _mRequestedGeomType_ would be the type I chose when creating the layer, right? And unless I edit xml manually it will be one of the types which actually exist in the table (at the time of adding a layer).
  When I'm overwriting _mDetectedType_ with _mRequestedGeomType_ I simply say "let's ASSUME the table still has the same geometry type". That is I assume table contents has not changed in principle (then I move a project to production environment that should always be the case).
  Therefore unless table contents have changed considerably, _mRequestedGeomType_ and _mDetectedType_ should be the same?

  Another example with polygon vs multipolygon. Say I chose polygon, so requestedtype is polygon. Now if I have estimated data on and the first fetched geometry from mixed type table is multipolygon - detectedtype would be multipolygon, should QGIS really start reshuffling configuration? That multipolygon could be one record in a million polygon table... In this case re-checking would be not only unnecessary, but even unwanted/harmful? Or am I missing something?
(You can enable row movement in Oracle and then rows could be moved/rowid changed impacting which are returned as first row even in the same db session)

P.S. The code you link to uses _mRequestedType_, so my proposed change should not have impact on that?

Reply:
> That is I assume table contents has not changed in principle 

You cannot be sure, you could have added any geometry type to your table and you want to filter only the type you select at first. If detectedGeomType == requestedGeomType, [filtering](https://github.com/qgis/QGIS/blob/master/src/providers/oracle/qgsoraclefeatureiterator.cpp#L174) is not possible anymore.

> the first fetched geometry from mixed type table is multipolygon - detectedtype would be multipolygon

Right! that's why I told you to try the GROUP BY. But I can't see a better solution when you use metadata

> P.S. The code you link to uses mRequestedType, so my proposed change should not have impact on that?

You mean mRequestedGeomType, the impact will be the one described above, no filtering.

Reply:
> If detectedGeomType == requestedGeomType, [filtering](https://github.com/qgis/QGIS/blob/master/src/providers/oracle/qgsoraclefeatureiterator.cpp#L174) is not possible anymore.

  Either I do not understand the code, or it is not correct.
  Say my requested geometry type is A.
  Table has rows with types A, B and C.
  When I (re)open the project, provider will check the data and get A, B or C randomly (could change on every load as oracle reshuffles data rows).
  How can this randomly selected type be used?
  And why, if I explicitly specified geometry type I want when creating vector layer?

  If it is really really needed to filter on geometry type, geometry type condition should be added **always** (remove _if_ condition at line 174) as rows (and therefore geometry types) could change **during** active session. QGIS Desktop could be opened for days or weeks, the same for QGIS Server.

Reply:
Ok, I had another look and now get the idea. Geometry type check puts geometry type into _mDetectedGeomType_, but if there is more than one geometry type found - then it puts a value of "unknown".
Then if _mRequestedGeometryType_ is not equal to detected one or if multiple types are detected then additional WHERE clause is added to filter on requested geometry type as well.

What else we have:
* as you already mentioned, this will not work when estimated data is used and therefore only 100 rows are checked (results would be inconsistent and unreliable).
* such checking during the opening of the project or creating a layer shows situation at that point in time and geometries (and new types) could be added after that.

Therefore:
* if somebody wants to have only requested geometry type, then such filter should be added **always**.
* such check will probably have some minor impact on performance, so probably adding this filtering of geometry type should be an option on layer level, or at least project level?

Reply:
We could maybe skip the detection of geom type if estimated meta data is true and the _mRequestedGeomType_ is set. And let the _mDetectedGeomType_ to Unknow. So the filter on geometry will be enabled from featureiterator.

But we can't force _mHasSpatialIndex_ and disable _mSrid_ detection.

It should not break anything, hoping that I don't miss a corner case.

Does it sounds good to you?

Reply:
I agree on skipping geom type testing when estimated meta data is true.

But SRID would have already been detected upon adding the layer. That is why I have a condition mSrid > 0. Why re-query?

Spatial index might be a separate issue (I'm ok with removing it from this PR and opening a new one specially for spacial index). The reason I'm setting it to true is that at least when using a **view** metadata shows there is no spatial index (as views do not have any indexes on them), but there actually is a spatial view on geometry table used in that view.
This index thing should probably be assumed to true when we're using a view, not a table.
Otherwise some other non trivial queries are done in loadField function which again slows down opening of project in QGIS Server (actually further/future idea is to skip loadField on project re-opening as well, as all fields are already saved in project file).

Reply:
> But SRID would have already been detected upon adding the layer. That is why I have a condition mSrid > 0. Why re-query?

Because it could be different from the one requested. When you add a layer to a project, you just add a configuration that point on a table, QGIS needs to check these parameters to be sure everything is fine and avoid issues later.

This is the way all providers are designed and I'm pretty sure that skipping all of this checks because of loading time won't be accepted and merged. 

IMHO I think you should focus more on how to speed up the different request behind this checks instead of trying to modify QGIS. But maybe other developers won't feel the same.

Reply:
SRID is fetched from metadata, so it's value will not change unless somebody actually decides to change data schema on-line. Therefore in normally designed system this should never happen.
I would say that QGIS should read all required metadata upon adding the layer, save it in the project and then simply use it (or refresh upon explicit request by the user). The only queries done on QGIS Server should be feature fetch queries. This would also increase all benchmarks of QGIS Server ;-)
But of course I'm not in a position to manage/rule here.
Therefore ok, I will only leave a code to skip geometry re-checking and will try to formulate further improvements in a separate PR (or should I first ask in developers mailing list?).
Thank you for your help!

Reply:
I think that if you are willing to skip all check at layer loading (reading the project), you should ask on developers mailing list. I'm pretty sure people won't agree with this, but it's always good to have others opinion.

> Thank you for your help!

You're welcome